pkg/controller/upgradeconfig/cluster_upgrader.go

```diff
@@ -0,0 +1,768 @@
+				readyRs = readyRs + 1
+			}
+
+		}
+	}
+
+	if totalRs != readyRs {
+		reqLogger.Info(fmt.Sprintf("not all replicaset are ready:expected number :%v , ready number %v", len(replicaSetList.Items), readyRs))
+		return false, nil
+	}
+
+	dsList := &appsv1.DaemonSetList{}
+	err = c.List(context.TODO(), dsList)
+	if err != nil {
+		return false, err
+	}
+	readyDS := 0
+	totalDS := 0
+	for _, ds := range dsList.Items {
+		if strings.HasPrefix(ds.Namespace, "default") ||
+			strings.HasPrefix(ds.Namespace, "kube") ||
+			strings.HasPrefix(ds.Namespace, "openshift") {
+			totalDS = totalDS + 1
+			if ds.Status.DesiredNumberScheduled == ds.Status.NumberReady {
+				readyDS = readyDS + 1
+			}
+		}
+	}
+	if len(dsList.Items) != readyDS {
+		reqLogger.Info(fmt.Sprintf("not all daemonset are ready:expected number :%v , ready number %v", len(dsList.Items), readyDS))
+		return false, nil
+	}
+
+	return true, nil
+}
+
+// TODO
+func RemoveMaintWindow(c client.Client, upgradeConfig *upgradev1alpha1.UpgradeConfig, reqLogger logr.Logger) (bool, error) {
+
+	return true, nil
+}
+
+// This perform cluster health check after upgrade
+func PostClusterHealthCheck(c client.Client, upgradeConfig *upgradev1alpha1.UpgradeConfig, reqLogger logr.Logger) (bool, error) {
+
+	return performClusterHealthCheck(c)
+}
+
+// Check whether nodes are upgraded or not
+func NodesUpgraded(c client.Client, nodeType string, reqLogger logr.Logger) (bool, error) {
+	configPool := &machineconfigapi.MachineConfigPool{}
+	err := c.Get(context.TODO(), types.NamespacedName{Name: nodeType}, configPool)
+	if err != nil {
+		return false, nil
+	}
+	//TODO send timeout alert if wait timeout
+	if configPool.Status.ReadyMachineCount != configPool.Status.UpdatedMachineCount {
+		errMsg := fmt.Sprintf("not all %s are upgraded, upgraded: %v, ready: %v", nodeType, configPool.Status.UpdatedMachineCount, configPool.Status.ReadyMachineCount)
+		reqLogger.Info(errMsg)
+		return false, fmt.Errorf(errMsg)
+	}
+
+	return true, nil
+}
+
+// This check whether control plane is upgraded or not
+func ControlPlaneUpgraded(c client.Client, upgradeConfig *upgradev1alpha1.UpgradeConfig, reqLogger logr.Logger) (bool, error) {
+
+	clusterVersion := &configv1.ClusterVersion{}
+
+	err := c.Get(context.TODO(), types.NamespacedName{Name: "version"}, clusterVersion)
+	if err != nil {
+		return false, err
+	}
+	for _, c := range clusterVersion.Status.History {
+		if c.State == configv1.CompletedUpdate && c.Version == upgradeConfig.Spec.Desired.Version {
+			return true, nil
+		}
+	}
+	return false, nil
+
+}
+
+// This trigger the upgrade process
+func (c ClusterUpgrader) UpgradeCluster(cl client.Client, upgradeConfig *upgradev1alpha1.UpgradeConfig, reqLogger logr.Logger) error {
+
+	log.Info("upgrading cluster")
+	history := upgradeConfig.Status.History.GetHistory(upgradeConfig.Spec.Desired.Version)
+	conditions := history.Conditions
+
+	history.Phase = upgradev1alpha1.UpgradePhaseUpgrading
```

Comment:
For correctness should this be set within `CommenceUpgrade` func?

Reply:
@sedroche I don't have a good answer here, could you create a separate issue discuss this? in my mind, if the process already started, it's `upgrading` in our own upgrade process.

Reply:
Ok gotcha. This is the difference between our upgradeconfig reconciling and the actual upgrade. Its not a big deal so just ignore if you want but is `Pending` a better fit here?